### PR TITLE
DROOLS-2374: [DMN Designer] Literal Expression Cell can not be updated once erased

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-2374

### CODE REVIEW

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/relation/RelationColumn.java [hunk 1]
```diff
@@ -95,15 +94,13 @@ public void edit(final GridCell<String> cell,
                      final GridBodyCellRenderContext context,
                      final Callback<GridCellValue<String>> callback) {
         factory.attachDomElement(context,
-                                 (e) -> e.getWidget().setValue(assertCell(cell).getValue().getValue()),
+                                 (e) -> e.getWidget().setValue(assertCellValue(assertCell(cell).getValue()).getValue()),
                                  (e) -> e.getWidget().setFocus(true));
     }
 
-    private GridCell<String> assertCell(final GridCell<String> cell) {
-        if (cell != null && cell.getValue() != null && cell.getValue().getValue() != null) {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/relation/RelationColumn.java]: We have replaced this line with:
```java
(e) -> e.getWidget().setValue(assertCellValue(assertCell(cell).getValue()).getValue())
```
The new implementation doesn't check if `cell.getValue().getValue() != null`. Is that correct?

In other words, if I add the test below into `BaseDOMElementSingletonColumnTest` the test will fail.
```java
@Test
    @SuppressWarnings("unchecked")
    public void checkEditNullCellValueValue() {
        final GridCell<String> cell = new BaseGridCell<>(new BaseGridCellValue<>(null));

        column.edit(cell,
                    context,
                    result -> {/*Nothing*/});

        assertCellEdit(DEFAULT_VALUE);
    }
```

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/model/DMNGridColumn.java [hunk 2]
```diff
@@ -51,6 +54,24 @@ public DMNGridColumn(final List<HeaderMetaData> headerMetaData,
...
+
+    protected GridCellValue<T> assertCellValue(final GridCellValue<T> cellValue) {
+        if (cellValue != null) {
+            return cellValue;
+        }
+        return makeDefaultCellValue();
+    }
+
+    protected GridCellValue<T> makeDefaultCellValue() {
+        return null;
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/model/DMNGridColumn.java]: Question here, wouldn't be better to throw an exception or make the method abstract to force user override appropriately in `DMNGridColumn` implemetations?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/relation/RelationColumn.java (hunk 1 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/relation/RelationColumn.java]: Good spot! Added check and your test. Thanks.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/model/DMNGridColumn.java (hunk 2 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/grid/model/DMNGridColumn.java]: I've moved the assertions to a new sub-class ```DMNSimpleGridColumn``` that is the base for all "simple" columns that use ```TextBox``` or ```TextArea```. The only columns not extending this are ```ExpressionEditorColumn```, ```UndefinedExpressionColumn``` (both of which provide a custom widget for editing the cell value) and ```function.supplementary.NameColumn``` that is not editable. There is now no need to either have all columns implement ```makeDefaultCellValue()``` or throw an exception.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
```
org.kie.workbench.common.services.backend.compiler.plugin.KieMetadataTest.compileAndLoadKieJarMetadataAllResourcesPackagedJar

Failing for the past 11 builds (Since Unstable#3113 )
Took 51 sec.
Stacktrace
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at org.kie.workbench.common.services.backend.compiler.plugin.KieMetadataTest.compileAndLoadKieJarMetadataAllResourcesPackagedJar(KieMetadataTest.java:109)
```

**Comment by manstis** [comment]
@ederign Test ```KieMetadataTest.compileAndLoadKieJarMetadataAllResourcesPackagedJar()``` has been failing for the last 11 builds. Could you please take a look (you made the last change to the class by the look of it).

**Comment by ederign** [comment]
@manstis I just touched this in project oriented refactoring! :-)

But anyway: https://github.com/kiegroup/kie-wb-common/pull/1498

**Comment by manstis** [comment]
Oh sorry @ederign but it did look like you knew why it was failing; so I thank-you :-)

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Jenkins please retest this.

**Pull request merged**